File: solutions/typescript/anagram/1/anagram.ts
export class Anagram {
  private target: string;
  private targetLower: string;
  private targetSorted: string;

  constructor(input: string) {
    this.target = input;
    this.targetLower = input.toLowerCase();
    this.targetSorted = this.sortString(this.targetLower);
  }

  private sortString(str: string): string {
    return str.split('').sort().join('');
  }

  public matches(...potentials: string[]): string[] {
    const results: string[] = [];
    
    for (const candidate of potentials) {
      // Skip if the candidate is the exact same word (case-sensitive)
      if (candidate === this.target) {
        continue;
      }
      
      const candidateLower = candidate.toLowerCase();
      
      // Skip if lengths don't match
      if (candidateLower.length !== this.targetLower.length) {
        continue;
      }
      
      // Skip if it's the same word (case-insensitive)
      if (candidateLower === this.targetLower) {
        continue;
      }
      
      // Check if sorted strings match
      if (this.sortString(candidateLower) === this.targetSorted) {
        results.push(candidate);
      }
    }
    
    return results;
  }
}

File: solutions/java/anagram/1/src/main/java/Anagram.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Anagram {
    private final String lowerCaseWord;
    private final char[] sortedWordChars;

    public Anagram(String word) {
        // Normalize the target word to lowercase for case-insensitive comparison
        this.lowerCaseWord = word.toLowerCase();
        
        // Sort the characters of the target word to easily check for anagrams later
        char[] chars = lowerCaseWord.toCharArray();
        Arrays.sort(chars);
        this.sortedWordChars = chars;
    }

    public List<String> match(List<String> candidates) {
        List<String> anagrams = new ArrayList<>();
        
        for (String candidate : candidates) {
            String lowerCaseCandidate = candidate.toLowerCase();
            
            // Rule: A word is not its own anagram (case-insensitive)
            if (lowerCaseCandidate.equals(lowerCaseWord)) {
                continue;
            }
            
            // Sort the candidate's characters to compare with the target
            char[] candidateChars = lowerCaseCandidate.toCharArray();
            Arrays.sort(candidateChars);
            
            // If the sorted characters match, it's an anagram!
            if (Arrays.equals(sortedWordChars, candidateChars)) {
                // Add the ORIGINAL candidate to preserve its original casing
                anagrams.add(candidate); 
            }
        }
        
        return anagrams;
    }
}